File: 1680_Concatenation_of_Consecutive_Binary_Numbers/solution.c
#define M 1000000007
int concatenatedBinary(int n) {
    unsigned long long result = 0;
    for (unsigned int i = 1; i <= (unsigned int)n; i++) {
        int b = 32 - __builtin_clz(i);
        result = (result << b) + i;
        if (result >= M) result %= M;
    }
    return (int)result;
}

File: 1680_Concatenation_of_Consecutive_Binary_Numbers/solution.py
class Solution:
    def concatenatedBinary(self, n: int) -> int:
        M = 10**9 + 7
        result = 0
        for i in range(1, n + 1):
            b = i.bit_length()
            result = (result << b) + i
            if result >= M: result %= M
        return result